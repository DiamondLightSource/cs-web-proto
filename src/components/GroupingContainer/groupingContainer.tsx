import React from "react";
import PropTypes from "prop-types";

import {
  Component,
  Widget,
  WidgetPropType,
  InferWidgetProps
} from "../Widget/widget";

const GroupingContainerProps = {
  name: PropTypes.string.isRequired,
  children: PropTypes.node
};

// Generic display widget to put other things inside
export const GroupingContainerComponent = (
  props: InferWidgetProps<typeof GroupingContainerProps> & Component
): JSX.Element => (
  // Uses an inner margin for children similar to Phoebus
  // This prevents the title being overwritten
  // Could be changed or perhaps customisable as a prop
  <fieldset
    style={{
      boxSizing: "border-box",
      ...props.style
    }}
  >
    <legend>{props.name}</legend>
    <div style={{ position: "absolute", top: "20px", left: "20px" }}>
      {props.children}
    </div>
  </fieldset>
);

const GroupingWidgetProps = {
  ...GroupingContainerProps,
  ...WidgetPropType
};

export const GroupingContainer = (
  props: InferWidgetProps<typeof GroupingWidgetProps>
): JSX.Element => <Widget baseWidget={GroupingContainerComponent} {...props} />;

GroupingContainer.propTypes = GroupingWidgetProps;
