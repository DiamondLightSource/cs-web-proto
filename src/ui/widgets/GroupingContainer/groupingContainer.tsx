import React from "react";

import { Component, Widget, WidgetPropType } from "../widget";
import { registerWidget } from "../register";
import { StringProp, ChildrenPropOpt, InferWidgetProps } from "../propTypes";

const GroupingContainerProps = {
  name: StringProp,
  children: ChildrenPropOpt
};

// Generic display widget to put other things inside
export const GroupingContainerComponent = (
  props: InferWidgetProps<typeof GroupingContainerProps> & Component
): JSX.Element => (
  // Uses an inner margin for children similar to Phoebus
  // This prevents the title being overwritten
  // Could be changed or perhaps customisable as a prop
  <fieldset
    style={{
      boxSizing: "border-box",
      ...props.style
    }}
  >
    <legend>{props.name}</legend>
    <div style={{ position: "relative", height: "100%", width: "100%" }}>
      {props.children}
    </div>
  </fieldset>
);

const GroupingWidgetProps = {
  ...GroupingContainerProps,
  ...WidgetPropType
};

export const GroupingContainer = (
  props: InferWidgetProps<typeof GroupingWidgetProps>
): JSX.Element => <Widget baseWidget={GroupingContainerComponent} {...props} />;

registerWidget(GroupingContainer, GroupingWidgetProps, "grouping");
