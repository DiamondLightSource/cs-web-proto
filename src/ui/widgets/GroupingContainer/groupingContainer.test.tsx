import React from "react";
import { GroupingContainerComponent } from "./groupingContainer";
import { shallow, ShallowWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";

let snapshot: ReactTestRenderer;
let wrapper: ShallowWrapper;

beforeEach((): void => {
  const grouping = <GroupingContainerComponent name={"Test"} />;
  wrapper = shallow(grouping);
  snapshot = create(grouping);
});

describe("<GroupingContainerComponent />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it is a div HTML object", (): void => {
    expect(wrapper.type()).toEqual("div");
  });

  test("it renders child div with text", (): void => {
    const childText = "Testing Child Component";
    const groupingWithChild = (
      <GroupingContainerComponent name={"Test"}>
        <div>{childText}</div>
      </GroupingContainerComponent>
    );
    const wrapperWithChild = shallow(groupingWithChild);
    expect(wrapperWithChild.childAt(0).text()).toEqual(childText);
  });
});
