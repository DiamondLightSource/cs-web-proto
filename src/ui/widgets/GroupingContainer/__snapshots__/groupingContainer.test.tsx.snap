// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupingContainerComponent /> it matches the snapshot 1`] = `
<div
  style={
    Object {
      "backgroundColor": undefined,
      "color": undefined,
      "outline": undefined,
      "outlineOffset": undefined,
      "visibility": undefined,
    }
  }
/>
`;
